File: packages/cli/src/commands/add.ts
import { isContainerRunning } from "../lib/docker";
import { log, postgresEnv, redisEnv, rabbitmqEnv, aistorEnv, signozEnv } from "../lib/output";
import { loadRegistry, saveRegistry } from "../lib/registry";
import { buildAddTargets } from "../lib/targets";
import type { ServiceName } from "../lib/types";
import { validateAppName } from "../lib/validate";
import * as postgres from "../services/postgres";
import * as redis from "../services/redis";
import * as rabbitmq from "../services/rabbitmq";
import * as aistor from "../services/aistor";
import * as signoz from "../services/signoz";

const SERVICE_CONTAINERS: Record<ServiceName, string> = {
  postgres: "infra-postgres",
  redis: "infra-redis",
  rabbitmq: "infra-rabbitmq",
  aistor: "infra-aistor",
  signoz: "infra-signoz-otel-collector",
};

const SERVICE_START_HINT: Partial<Record<ServiceName, string>> = {
  signoz: "make signoz-up",
};

export interface AddOptions {
  envs?: string[];
  env?: string;
}

export async function addCommand(
  appName: string,
  services: ServiceName[],
  options?: AddOptions,
): Promise<void> {
  if (!appName) {
    log.error("App name is required.");
    process.exit(1);
  }

  try {
    validateAppName(appName);
  } catch {
    log.error('Invalid app name. Use lowercase letters, numbers, and hyphens.');
    process.exit(1);
  }

  if (services.length === 0) {
    log.error("At least one service is required. Use -s aistor,...");
    process.exit(1);
  }

  const registry = await loadRegistry();

  const targets = buildAddTargets(appName, options, registry);

  for (const target of targets) {
    if (!registry.apps[target.name]) {
      log.error(`App "${target.name}" not found. Create it first with: bun shelf setup ${target.name} -s ...`);
      process.exit(1);
    }
  }

  // Containers are shared across targets — check once.
  for (const service of services) {
    const container = SERVICE_CONTAINERS[service];
    if (!(await isContainerRunning(container))) {
      const hint = SERVICE_START_HINT[service] ?? "make up";
      log.error(`Container "${container}" is not running. Start it with: ${hint}`);
      process.exit(1);
    }
  }

  for (const target of targets) {
    const app = registry.apps[target.name];
    const toProvision: ServiceName[] = [];
    for (const service of services) {
      if (app.services[service]) {
        log.warn(`${target.name}: ${service} already provisioned — skipping`);
      } else {
        toProvision.push(service);
      }
    }

    if (toProvision.length === 0) {
      log.info(`${target.name}: nothing to do.`);
      continue;
    }

    const results: string[] = [];

    for (const service of toProvision) {
      try {
        switch (service) {
          case "postgres": {
            const config = await postgres.provision(target.name);
            app.services.postgres = config;
            results.push(postgresEnv(config));
            break;
          }
          case "redis": {
            const config = await redis.provision(target.name);
            app.services.redis = config;
            results.push(redisEnv(config));
            break;
          }
          case "rabbitmq": {
            const config = await rabbitmq.provision(target.name);
            app.services.rabbitmq = config;
            results.push(rabbitmqEnv(config));
            break;
          }
          case "aistor": {
            const config = await aistor.provision(target.name);
            app.services.aistor = config;
            results.push(aistorEnv(config));
            break;
          }
          case "signoz": {
            const config = await signoz.provision(target.name, {
              serviceName: target.signozServiceName,
              environment: target.signozEnv,
            });
            app.services.signoz = config;
            results.push(signozEnv(config));
            break;
          }
        }
        log.success(`${target.name}: ${service} provisioned`);
      } catch (err) {
        log.error(`Failed to provision ${service} for ${target.name}: ${err instanceof Error ? err.message : err}`);
        process.exit(1);
      }
    }

    await saveRegistry(registry);

    console.log("");
    log.title(`Services attached to "${target.name}":\n`);
    console.log(results.join("\n\n"));
    console.log("");
  }
}

File: internal/shelfcore/add.go
package shelfcore

import (
	"context"
	"fmt"

	"github.com/IvanMicai/infra-shelf/internal/envspec"
	"github.com/IvanMicai/infra-shelf/internal/registry"
)

type AddOptions struct {
	Services []string
	Envs     []string
	Env      string
}

// AddServices attaches services to existing app(s). Services already
// provisioned on a target are skipped with a warning. Mirrors `shelf add`.
func (e *Engine) AddServices(ctx context.Context, appName string, opts AddOptions) ([]TargetResult, error) {
	if err := registry.ValidateAppName(appName); err != nil {
		return nil, err
	}
	if len(opts.Services) == 0 {
		return nil, ErrNoServices
	}
	if _, err := registry.ParseServices(opts.Services); err != nil {
		return nil, err
	}

	reg, err := e.Store.Load()
	if err != nil {
		return nil, err
	}

	targets := envspec.BuildAddTargets(appName, envspec.Options{Envs: opts.Envs, Env: opts.Env}, reg)

	for _, target := range targets {
		if _, ok := reg.Apps[target.Name]; !ok {
			return nil, fmt.Errorf("%w: %s", ErrAppNotFound, target.Name)
		}
	}

	if err := containerCheck(ctx, opts.Services); err != nil {
		return nil, err
	}

	results := make([]TargetResult, 0, len(targets))
	for _, target := range targets {
		entry := reg.Apps[target.Name]
		result := TargetResult{Name: target.Name, Failures: map[string]error{}}

		toProvision := []string{}
		for _, svc := range opts.Services {
			if hasService(entry, svc) {
				e.Reporter.Warn(fmt.Sprintf("%s: %s already provisioned — skipping", target.Name, svc))
				continue
			}
			toProvision = append(toProvision, svc)
		}

		if len(toProvision) == 0 {
			e.Reporter.Info(fmt.Sprintf("%s: nothing to do", target.Name))
			result.Entry = entry
			results = append(results, result)
			continue
		}

		provOpts := provisionOptions{
			SignozServiceName: target.SignozServiceName,
			SignozEnvironment: target.SignozEnv,
		}

		for _, svc := range toProvision {
			cfg, err := provisionService(ctx, svc, target.Name, provOpts)
			if err != nil {
				e.Reporter.Error(fmt.Sprintf("failed to provision %s for %s: %v", svc, target.Name, err))
				result.Failures[svc] = err
				continue
			}
			setServiceConfig(&entry, svc, cfg)
			result.Provisioned = append(result.Provisioned, svc)
			e.Reporter.Success(fmt.Sprintf("%s: %s provisioned", target.Name, svc))
		}

		reg.Apps[target.Name] = entry
		if err := e.Store.Save(reg); err != nil {
			return results, fmt.Errorf("save registry: %w", err)
		}
		result.Entry = entry
		results = append(results, result)
	}

	return results, nil
}
